File: core/src/main/java/io/github/shaksternano/borgar/core/command/EmojiImageCommand.java
package io.github.shaksternano.borgar.core.command;

import com.google.common.collect.ListMultimap;
import io.github.shaksternano.borgar.core.command.util.CommandResponse;
import io.github.shaksternano.borgar.core.util.MessageUtil;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;

import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * Gets the image of an emoji.
 */
public class EmojiImageCommand extends BaseCommand<Void> {

    /**
     * Creates a new command object.
     *
     * @param name        The name of the command. When a user sends a message starting with {@link Command#PREFIX}
     *                    followed by this name, the command will be executed.
     * @param description The description of the command. This is displayed in the help command.
     */
    public EmojiImageCommand(String name, String description) {
        super(name, description);
    }

    /**
     * Sends the image URL of the first emoji found in the message, the message it's replying to, or a previously sent message.
     *
     * @param arguments      The arguments of the command.
     * @param extraArguments A multimap mapping the additional parameter names to a list of the arguments.
     * @param event          The {@link MessageReceivedEvent} that triggered the command.
     * @return A {@code CompletableFuture} that completes with a list of {@code MessageCreateData} that will be sent to the channel where the command was triggered.
     */
    @Override
    public CompletableFuture<CommandResponse<Void>> execute(List<String> arguments, ListMultimap<String, String> extraArguments, MessageReceivedEvent event) {
        return MessageUtil.processMessages(event.getMessage(), MessageUtil::getFirstEmojiUrl)
            .thenApply(urlOptional ->
                new CommandResponse<>(urlOptional.orElse("No emoji found!"))
            );
    }
}

File: chat/src/main/kotlin/io/github/shaksternano/borgar/chat/command/EmojiImageCommand.kt
package io.github.shaksternano.borgar.chat.command

import io.github.shaksternano.borgar.chat.event.CommandEvent
import io.github.shaksternano.borgar.chat.util.getFirstEmojiUrl
import io.github.shaksternano.borgar.chat.util.searchOrThrow
import io.github.shaksternano.borgar.core.io.task.FileTask
import io.github.shaksternano.borgar.core.io.task.UrlFileTask

object EmojiImageCommand : FileCommand(
    CommandArgumentInfo(
        key = "emoji",
        description = "The emoji to get the image of.",
        type = CommandArgumentType.String,
        required = false,
    ),
    requireInput = false,
) {

    override val name: String = "emojiimage"
    override val aliases: Set<String> = setOf("emoji")
    override val description: String = "Gets the image of an emoji."

    override suspend fun createTask(arguments: CommandArguments, event: CommandEvent, maxFileSize: Long): FileTask {
        val emojiUrl = event.asMessageIntersection(arguments)
            .searchOrThrow(
                "No emojis found.",
                CommandMessageIntersection::getFirstEmojiUrl,
            )
        return UrlFileTask(emojiUrl)
    }
}